frontend/src/app/feature/room/room.component.ts

```diff
@@ -39,131 +48,126 @@ export class RoomComponent extends WithWebSocket implements OnInit, OnDestroy {
   }
 
   ngOnInit() {
-    this.route.paramMap.subscribe(params => {
-      const roomId = params.get('roomId');
-
-      if (this.dashboardService.roomIds.indexOf(roomId) === -1) {
-        this.leaveRoom();
-        return;
-      }
-
-      this.room = new Room(roomId);
-
-      const myPlayer = this.dashboardService.getPlayer(roomId);
-      if (myPlayer) {
-        this.init(myPlayer);
-      } else {
-        setTimeout(() => { // hack to avoid 'value has been changed before check'
-          this.dialog.open(RoomDialogComponent).afterClosed().subscribe(this.onDialogClose(roomId));
-        });
-      }
-    });
-  }
-
-  private init(player: Player) {
-    this.you = player;
-    this.room.addPlayer(this.you);
-    this.connect(`${WsConstants.roomTopic}${this.room.id}`);
-  }
-
-  private onDialogClose(roomId: string) {
-    return (name: string) => {
-      if (!name) {
-        this.leaveRoom();
-      } else {
-        this.init(new Player(this.userService.userId, name));
-        this.dashboardService.addPlayer(roomId, this.you);
-      }
-    };
+    this.route.paramMap.pipe(
+      map(params => params.get('roomId')),
+      flatMap(roomId => this.roomService.getRoom(roomId)),
+      tap(room => this.throwIfEmpty(room)),
+      tap(room => {
+        room.players.map(p => Player.addColorIfMissing(p));
```

Comment:
map creates new instance of collection, use forEach instead

Reply:
Right, fixed